---
title: Marker and Sample Diversity
author: Alfred Hubbard
output:
  html_notebook: default
  pdf_document: default
  github_document: default
---

```{r setup, echo = FALSE, message = FALSE, warning = FALSE}
# Load required libraries ----------------------------------------------
library(adegenet)
library(pegas)
library(poppr)
# These will be referenced without the `package::` construct, and thus 
# are loaded second to avoid masking
library(knitr)
library(magrittr)
library(tidyverse)

# Don't include code in PDF output
opts_chunk$set(echo = ! is_latex_output(), out.width = "100%")
# Set default ggplot2 theme
theme_set(theme_bw())

# Hardcode parameters. These are not passed as shell arguments because 
# hardcoded values facilitate development and this document is not 
# meant to be usable with other projects.
inp <- list(
  mh_hap = "../../results/microhap/MalariaGEN/mg_microhap_haplo.rds", 
  mh_gd = 
    "../../results/microhap/MalariaGEN/mg_microhap_filtered4popgen_gd.rds", 
  mh_csv = "../../results/microhap/MalariaGEN/mg_microhap.csv"
)
```

# Marker Diversity

Nucleotide diversity was computed for each marker with `pegas::nuc.div()`. Since
this was done with pegas `haplotype` objects, the frequencies of each haplotype
were used to perform this calculation.

Note that when identifying haplotypes, pegas treats gaps as separate characters 
(meaning a Hamming distance of 1 from any other character) but ambiguities are 
treated as having a Hamming distance of 0 from nucleotides included in that 
ambiguity. Even so, sometimes it is not possible to assign a sequence with an 
ambiguity at a given position to other haplotypes that lack ambiguities at that 
position. In this scenario, pegas assigns this sequence to a new haplotype. That 
means that there tend to be a few low frequency haplotypes for each marker that 
were assigned this way because of ambiguities, and would be merged with one of 
the other haplotypes if the ambiguity could be resolved.

All of that is to say the diversity numbers computed by pegas may be slightly
inflated. Relative comparisons should be valid, but absolute interpretation, or
richness statistics, should be avoided.

Note also that only gaps in the middle of the sequence are treated as separate
characters. Gaps at the end of the sequences are treated as Ns, and so will
generally be assigned to other haplotypes.

```{r fig.width = 10}
# Compute and visualize nucleotide diversity ---------------------------
nuc_div_res <- read_rds(inp$mh_hap) %>%
  # Compute nucleotide diversity, telling pegas to remove sites with 
  # missing data in a pairwise manner to preserve the maximum amount of 
  # data in each individual distance calculation
  mutate(nuc_div = map_dbl(hap, pegas::nuc.div, pairwise.deletion = TRUE)) %>%
  select(-hap)
nuc_div_res %>%
  ggplot(mapping = aes(x = target, y = nuc_div)) +
  geom_col() +
  labs(x = "Target", y = "Nucleotide Diversity") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

This bar plot shows the nucleotide diversity for each marker. There is
considerable variability in the diversity of these markers, and a number don't
have any segregating sites, leading to a diversity of 0.

# Genotype Accumulation Curve

A genotype accumulation curve was calculated to assess whether there are
sufficient markers in the dataset to distinguish individual samples.

First, loci with only one allele were filtered out of the dataset:

```{r}
mh_data_gd <- read_rds(inp$mh_gd) %>%
  poppr::informloci()
```

Then the curve was plotted:

```{r fig.width = 10}
poppr::genotype_curve(mh_data_gd, sample = 1000, quiet = TRUE)
```

From this figure, it is apparent that there are plenty of markers in the dataset
to distinguish different samples, even once loci with only one allele were
removed. In fact, it appears that only about 15 markers would be sufficient for
this purpose.

# Sample Diversity

To complement marker diversity, sample-level diversity was also estimated with
Nei's expected heterozygosity and evenness was calculated with the $E_5$ 
statistic. Two population strata were used for the entire dataset: populations 
as defined by the MalariaGEN project ("Population") and countries. In addition, 
diversity was calculated at the site level for Cambodia and Vietnam, to 
facilitate comparison with some of the results in relatedness.Rmd.

## By Population

```{r}
# Compute summary statistics, filtering to those of interest -----------
adegenet::setPop(mh_data_gd) <- ~Population
poppr::poppr(mh_data_gd) %>%
  select(Pop, N, MLG, Hexp, E.5)
```

This table gives the number of individuals, the number of multi-locus genotypes
(MLGs) observed in those individuals, Nei's expected heterozygosity, and 
evenness for each population.

Diversity is generally fairly low and evenness is high, although Latin America 
(LAM) has a lower evenness than other populations and Oceania (OCE) has a 
particularly low diversity. The high evenness is not a surprise given that the
number of MLGs is high relative to the number of individuals - most individuals
constitute a unique MLG, meaning the abundances of the MLGs tend to be low, and
therefore even.

## By Country

```{r}
# Compute summary statistics, filtering to those of interest -----------
adegenet::setPop(mh_data_gd) <- ~Country
poppr::poppr(mh_data_gd) %>%
  select(Pop, N, MLG, Hexp, E.5)
```

Indonesia and Colombia stand out as having lower diversity than the other
countries, and Colombia also stands out as having lower evenness. This makes
sense in light of the population-level results.

## By Site

```{r}
# Read metadata and make site/country key ------------------------------
site_country_key <- read_csv(
    inp$mh_csv, col_types = cols(
      .default = col_character(), 
      Lat = col_double(), 
      Long = col_double(), 
      Year = col_integer(), 
      `% callable` = col_double(), 
      Fws = col_double(), 
      F_MISS = col_double()
    ), 
    progress = FALSE
  ) %>%
  select(Site, Country) %>%
  distinct()

# Compute summary statistics, filtering to those of interest -----------
adegenet::setPop(mh_data_gd) <- ~Site
poppr::poppr(mh_data_gd) %>%
  left_join(site_country_key, by = c("Pop" = "Site")) %>%
  relocate(Country, .after = Pop) %>%
  filter(Country %in% c("Cambodia", "Vietnam")) %>%
  select(Pop, Country, N, MLG, Hexp, E.5)
```

As at the coarser levels of aggregation, we see that the sites in Cambodia and
Vietnam tend to have low diversity and high evenness.
