---
title: Linkage Disequilibrium Test for Microhaplotype Data
author: Alfred Hubbard
output:
  html_notebook: default
  pdf_document: default
  github_document: default
---

```{r setup, echo = FALSE, message = FALSE, warning = FALSE}
# Load required libraries ----------------------------------------------
library(hubpopgen)
library(pegas)
library(poppr)
# These will be referenced without the `package::` construct, and thus 
# are loaded second to avoid masking
library(dplyr)
library(ggplot2)
library(knitr)
library(magrittr)
library(purrr)
library(readr)
library(stringr)
library(tibble)
library(tidyr)

# Don't include code in PDF output
opts_chunk$set(echo = ! is_latex_output(), out.width = "100%")
# Set seed for reproducibility
set.seed(59404051)

# Hardcode parameters. These are not passed as shell arguments because 
# hardcoded values facilitate development and this document is not 
# meant to be usable with other projects.
inp <- list(
  microhap_gd = "../../results/microhap/microhap_filtered4ld_gd.rds", 
  microhap_tidy = "../../results/microhap/microhap_filtered4ld.csv", 
  ld_res_poppr = "../../results/ld_res_poppr.rds"
)
out <- list(trgs2filter = "../../results/trgs2filter/ld.csv")
```

# pegas

Linkage disequilibrium (LD) among locus pairs is estimated with the $T_2$ 
statistic described in Zaykin et al. (2008). This method is based on the 
correlation among allele frequencies, and "can be interpreted as the total 
correlation between a pair of loci."

This statistic was calculated among all pairs of loci, excluding those that are
on different chromosomes.

```{r warning = FALSE}
# Read in data and covert to the loci class ----------------------------
mh_data_gd <- read_rds(inp$microhap_gd)
mh_data_l <- pegas::genind2loci(mh_data_gd)

# Create tibble of locus name/number pairs -----------------------------
locus_nums <- attr(mh_data_l, "locicol")
locus_names <- names(mh_data_l)[locus_nums]
locus_num_name_key <- tibble(locus_name = locus_names, locus_num = locus_nums)
ld_res_pegas <- combn(locus_nums, 2) %>%
  t() %>%
  as_tibble() %>%
  rename(locus_a_num = V1, locus_b_num = V2) %>%
  left_join(locus_num_name_key, by = c("locus_a_num" = "locus_num")) %>%
  rename(locus_a_name = locus_name) %>%
  left_join(locus_num_name_key, by = c("locus_b_num" = "locus_num")) %>%
  rename(locus_b_name = locus_name)

# Read in microhaplotype data table with metadata ----------------------
mh_data <- read_csv(
  inp$microhap_tidy,  
  col_types = cols(
    .default = col_character(), 
    n_read = col_integer(), 
    n_samp_wdata = col_integer(), 
    n_trg_wdata = col_integer(), 
    ParasiteDensity = col_double(), 
    CT_Pv = col_double(), 
    CT_Pf18s = col_double(), 
    CT_PfvarATS = col_double(), 
    Age = col_integer(), 
    n_hap = col_integer(), 
    moi = col_integer()
  ), 
  progress = FALSE
)

# Join and filter by chromosome information ----------------------------
chrom_info <- mh_data %>%
  select(target, chrom) %>%
  distinct()
ld_res_pegas <- ld_res_pegas %>%
  left_join(chrom_info, by = c("locus_a_name" = "target")) %>%
  rename(chrom_a = chrom) %>%
  left_join(chrom_info, by = c("locus_b_name" = "target")) %>%
  rename(chrom_b = chrom) %>%
  # Regardless of what the test shows, we know that targets on 
  # different chromosomes cannot be physically linked
  filter(chrom_a == chrom_b) %>%
  select(-chrom_a, -chrom_b)

# Estimate LD ----------------------------------------------------------
ld_res_pegas <- ld_res_pegas %>%
  # Note that pegas prints a warning message about unphased genotypes 
  # when an individual is missing data at a given locus, regardless of 
  # whether the genotypes actually are unphased (or, indeed, regardless 
  # of whether the ploidy is even greater than 1)
  mutate(
    ld_res = map2(
      locus_a_num, 
      locus_b_num, 
      # Note that while in the locicol attribute the pop column is 
      # included in locus numbering (meaning the first locus column is 
      # two), this function wants the locus numbering to start at 1...
      ~pegas::LD(mh_data_l, locus = c(.x - 1, .y - 1))
    )
  ) %>%
  mutate(T2_pval = map_dbl(ld_res, pluck, "T2", "P-val")) %>%
  # Remove NA p-values, which indicate at least one of the loci only 
  # has one allele
  filter(! is.na(T2_pval))

# Visualize distribution of p-values -----------------------------------
ld_res_pegas %>%
  ggplot(mapping = aes(x = T2_pval)) +
  geom_histogram(bins = 30) +
  labs(x = "T_2 Test p-value", y = "No. of Locus Pairs")
```

The distribution of *p*-values is given above. It is clearly not uniform, 
suggesting some linkage is present in the dataset.

The Bonferroni-adjusted threshold for significance at the 0.05 level is:

```{r}
bf_thres <- 0.05 / nrow(ld_res_pegas)
bf_thres
```

The loci pairs that have significant LD at this threshold are:

```{r}
ld_res_pegas_signif <- ld_res_pegas %>%
  filter(T2_pval < bf_thres)
ld_res_pegas_signif %>%
  select(-locus_a_num, -locus_b_num, -ld_res)
```

Below is detailed information on the test results for these pairs.

```{r}
for (i in seq(1:nrow(ld_res_pegas_signif))) {
  print(
    str_c(
      "Comparison between ", 
      ld_res_pegas_signif$locus_a_name[[i]], 
      " and ", 
      ld_res_pegas_signif$locus_b_name[[i]]
    )
  )
  print(ld_res_pegas_signif$ld_res[[i]])
}
```

To create a dataset suitable for relatedness analysis, loci will be removed to
yield a filtered dataset without loci pairs that are in significant LD.

The two pairs on chromosome 4 share a locus in common:
PvP01_04_v1_401301_401500. Similarly, the two pairs on chromosome 12 also share
a locus in common: PvP01_12_v1_2032801_2033000. It is therefore logical to
remove these two loci.

The two pairs on chromosome 9 do not share a locus in common. Instead, the
sample size for each target is used to help decide which loci to discard:

```{r}
# Compute sample size for each target of interest ----------------------
chrom9_signif_loci <- ld_res_pegas_signif %>%
  # Because everything should be on one chromosome, filtering by 
  # target_b as well should be unnecessary
  filter(str_detect(locus_a_name, "PvP01_09")) %>%
  select(locus_a_name, locus_b_name) %>%
  pivot_longer(everything(), names_to = "pair_loc", values_to = "target") %$%
  target
mh_data %>%
  group_by(target) %>%
  summarize(n_samp = n_distinct(sample_id), .groups = "drop") %>%
  filter(target %in% chrom9_signif_loci)
```

To maintain the largest amount of data possible after filtering, we choose to
remove PvP01_09_v1_1564401_1564600 and PvP01_09_v1_769601_769800.

```{r}
tibble(
    target = c(
      "PvP01_04_v1_401301_401500", 
      "PvP01_12_v1_2032801_2033000", 
      "PvP01_09_v1_1564401_1564600", 
      "PvP01_09_v1_769601_769800"
    )
  ) %>%
  write_csv(out$trgs2filter)
```

# poppr

## Overall

Linkage disequilibrium was also tested using the $\bar{r}_{d}$ measure. This was 
used in favor of the more common index of association because the index of 
association is a ratio of the variance observed between individuals at each 
locus and the variance expected under conditions of linkage equilibrium. This 
measure increases with the number of loci, so $\bar{r}_{d}$ was developed as an
alternative that approximates the index of association while avoiding this
limitation.

Having multiple copies of a given multi-locus genotype can bias the estimation
of LD. To avoid this bias, these duplicate genotypes were removed using 
`poppr::clonecorrect`. LD was then estimated using `poppr::ia`.

```{r}
ld_res_poppr <- read_rds(inp$ld_res_poppr)
ld_res_poppr
```

Higher values of $\bar{r}_{d}$ indicate higher levels of linkage disequilibrium. 
Note that this value of overall LD includes estimates of LD between pairs of 
loci on different chromosomes, and thus should be interpreted with caution.

## Pairwise

To understand which pairs of loci are responsible for this LD, $\bar{r}_{d}$ was 
calculated and plotted separately for each pair of loci.

```{r fig.height = 12, fig.width = 14}
pairwise_ld_res <- poppr::clonecorrect(mh_data_gd, strata = NA) %>%
  poppr::pair.ia()
```

Targets with gray swatches only have one allele in the dataset, and so 
$\bar{r}_{d}$ cannot be calculated.

This method shows some agreement with the $T_2$ statistic, but most of the 
highest LD pairs are located on different chromosomes. We selected the $T_2$ 
results as our method of choice for subsequent filtering.
