---
title: Tidying of Microhaplotype Data from UCI 12/23 Run
author: Alfred Hubbard
output:
  html_notebook: default
  pdf_document: default
  github_document: default
---

```{r setup, echo = FALSE, message = FALSE, warning = FALSE}
# Load required libraries ----------------------------------------------
# These will be referenced without the `package::` construct, and thus 
# are loaded second to avoid masking
library(knitr)
library(magrittr)
library(readxl)
library(tidyverse)

# Don't include code in PDF output
opts_chunk$set(echo = ! is_latex_output())
# Set default ggplot2 theme
theme_set(theme_bw())

# Hardcode parameters. These are not passed as shell arguments because 
# hardcoded values facilitate development and this document is not 
# meant to be usable with other projects.
inp <- list(
  cigar_table = "../../results/AmpSeq/uci1223/asv2cigar/seqtab_cigar.tsv", 
  locus_metadata = "../../results/locus_metadata.tsv", 
  rep_metadata = "../../resources/MH_template__LB_EHS_dz07242024_to_liz.xlsx"
)
out <- list(
  allele_table = "../../results/microhap/uci1223/uci1223_microhap.tsv"
)
```

# Missing Data By Sample

For each sample, the replicate with the most loci that have 100 reads or more is
kept. At the same time, any samples where no replicates have 100 reads or more
are removed.

```{r}
# Read in final, filtered data -----------------------------------------
allele_table <- read_tsv(
    inp$cigar_table, 
    col_types = cols(.default = col_integer(), sample = col_character()), 
    progress = FALSE
  ) %>%
  pivot_longer(-sample, names_to = "locus_cigar", values_to = "n_read") %>%
  separate_wider_delim(locus_cigar, ",", names = c("locus_pos", "cigar")) %>%
  rename(rep_id = sample) %>%
  # Remove explicit missing data
  filter(n_read > 0)

# Read and join locus names --------------------------------------------
locus_metadata <- read_tsv(
    inp$locus_metadata, 
    col_types = cols(
      .default = col_character(), 
      start_pos = col_integer(), 
      end_pos = col_integer()
    ), 
    progress = FALSE
  ) %>%
  select(locus, chrom, start_pos, end_pos) %>%
  unite(start_end, start_pos, end_pos, sep = "-") %>%
  unite(locus_pos, chrom, start_end, sep = ":")
allele_table <- allele_table %>%
  left_join(locus_metadata, by = "locus_pos") %>%
  select(-locus_pos)

# Read and join replicate metadata -------------------------------------
rep_metadata <- read_xlsx(
    inp$rep_metadata, 
    sheet = "pv242"
  ) %>%
  select(SampleID, MH_DNA_Well) %>%
  rename(rep_id = SampleID, sample_id = MH_DNA_Well)
allele_table <- allele_table %>%
  left_join(rep_metadata, by = "rep_id")

# Filter out non-diversity loci ----------------------------------------
allele_table <- allele_table %>%
  filter(str_detect(locus, "PvP01"))

# For each sample, keep replicate with most loci with >= 100 reads -----
reps_w_highreads <- allele_table %>%
  group_by(rep_id, locus, sample_id) %>%
  summarize(rl_total_reads = sum(n_read), .groups = "drop") %>%
  filter(rl_total_reads >= 100) %>%
  group_by(rep_id, sample_id) %>%
  summarize(n_loc_highreads = n(), .groups = "drop") %>%
  group_by(sample_id) %>%
  filter(n_loc_highreads == max(n_loc_highreads)) %>%
  ungroup()
allele_table <- allele_table %>%
  filter(rep_id %in% reps_w_highreads$rep_id) %>%
  select(-rep_id)

# Compute number of loci with data for each sample ---------------------
n_loc_wdata <- allele_table %>%
  group_by(locus, sample_id) %>%
  summarize(n_read = sum(n_read), .groups = "drop") %>%
  group_by(sample_id) %>%
  summarize(n_loc_wdata = sum(n_read > 0), .groups = "drop")
allele_table <- allele_table %>%
  left_join(n_loc_wdata, by = "sample_id")

# Plot distribution of missing data proportion by locus ----------------
n_loc_wdata_thres <- 40
allele_table %>%
  select(sample_id, n_loc_wdata) %>%
  distinct() %>%
  ggplot(mapping = aes(x = n_loc_wdata)) +
  geom_histogram(bins = 20) +
  geom_vline(xintercept = n_loc_wdata_thres, color = "blue") +
  labs(x = "No. of Loci With Data", y = "No. of Samples")
```

This histogram shows the distribution of samples according to how many loci have 
data for each sample. The vertical blue line shows the threshold of `r 
n_loc_wdata_thres` loci with data that will be used for filtering the samples.

```{r}
# Filter out samples with lots of missing data -------------------------
allele_table <- allele_table %>%
  filter(n_loc_wdata >= n_loc_wdata_thres) %>%
  select(-n_loc_wdata)
```

# Missing Data by Locus

```{r}
# Compute number of samples with data for each loci --------------------
n_samp_wdata <- allele_table %>%
  group_by(locus, sample_id) %>%
  summarize(n_read = sum(n_read), .groups = "drop") %>%
  group_by(locus) %>%
  summarize(n_samp_wdata = sum(n_read > 0), .groups = "drop")
allele_table <- allele_table %>%
  left_join(n_samp_wdata, by = "locus")

# Plot data quantity by locus ------------------------------------------
n_samp_wdata_thres <- 20
allele_table %>%
  select(locus, n_samp_wdata) %>%
  distinct() %>%
  ggplot(mapping = aes(x = n_samp_wdata)) +
  geom_histogram(bins = 25) +
  geom_vline(xintercept = n_samp_wdata_thres, color = "blue") +
  labs(x = "No. of Samples With Data", y = "No. of Loci")
```

This histogram shows the distribution of loci according to how many samples have 
data for each locus. Note the sample filter of `r n_loc_wdata_thres` loci with 
data has already been applied, and this histogram is based on the remaining 
data.

The vertical blue line marks `r n_samp_wdata_thres` samples with data. This 
threshold is used to filter loci for subsequent analyses, removing those that 
fall below.

```{r}
# Filter out loci with lots of missing data ----------------------------
allele_table <- allele_table %>%
  filter(n_samp_wdata >= n_samp_wdata_thres) %>%
  select(-n_samp_wdata)
```

# Stats

Number of population genetics loci in the filtered dataset:

```{r}
allele_table %$%
  n_distinct(locus)
```

Number of samples in the filtered dataset:

```{r}
n_distinct(allele_table$sample_id)
```

```{r}
# Save filtered data
allele_table %>%
  # Rename to column names expected by PGEcore tools
  rename(
    specimen_id = sample_id, 
    target_id = locus, 
    # This is not a sequence, obviously, but that won't matter for the 
    # tools we are using
    seq = cigar, 
    read_count = n_read
  ) %>%
  write_tsv(out$allele_table)
```
